### PR TITLE
Semaphore implemented for FreeRTOS thread library

### PR DESCRIPTION
**Pull request by forGGe**

### CODE REVIEW

**Comment by forGGe** [comment]
Described in #11

File: lib/thread/freertos/export/ecl/thread/semaphore.hpp [hunk 1]
```diff
@@ -6,28 +6,32 @@
...
 
-//!
 //! \brief FreeRTOS-based semaphore.
-//! \TODO: notes about FreeRTOS semaphore limitation
+//! This semaphore is specific. It can only be raisen or cleared.
+//! Default state is cleared, so calling wait() would block until someone
+//! will call signal()
 //!
-class semaphore
+class binary_semaphore
```
Comment (rasendubi) [inline lib/thread/freertos/export/ecl/thread/semaphore.hpp]: Binary semaphore is usually called mutex. Don't you like that naming?

File: lib/thread/freertos/semaphore.cpp [hunk 2]
```diff
@@ -5,45 +5,41 @@
...
+        m_semaphore.wait();
+    }
 }
 
 ecl::err ecl::semaphore::try_wait()
 {
-    auto rc = xSemaphoreTake(m_semaphore, 0);
-    return rc == pdTRUE ? ecl::err::ok : ecl::err::again;
+    int saved;
+    if ((saved = m_cnt.fetch_sub(1)) >= 0) {
```
Comment (rasendubi) [inline lib/thread/freertos/semaphore.cpp]: Why not
```c
int saved = m_cnt.fetch_sub(1);
if (saved >= 0) {
  ...
```

File: lib/thread/freertos/export/ecl/thread/semaphore.hpp (hunk 1 above)
Comment (forGGe) [inline lib/thread/freertos/export/ecl/thread/semaphore.hpp]: Mutex is something that must be locked and unlocked within the same thread, effectively creating critical section. The Core already has one: 
https://github.com/forGGe/theCore/blob/develop/lib/thread/freertos/mutex.cpp#L5
Ref:
https://github.com/cjlano/freertos/blob/master/FreeRTOS/Source/include/semphr.h#L722

**Comment by rasendubi** [comment]
Don't see an implementation of the binary_semaphore. Did you forget to commit it?

File: lib/thread/freertos/semaphore.cpp [hunk 3]
```diff
@@ -5,45 +5,41 @@
...
+        return err::ok; // Got semaphore.
+    } else if (saved == -1) {
+        // Unblock any poor thread that was blocked just right now.
+        m_semaphore.signal();
```
Comment (rasendubi) [inline lib/thread/freertos/semaphore.cpp]: m_cnt is 0.

Thread 1:
- try_wait()
- fetch_sub return -1

Preemption

Thread 2:
- wait()
- fetch_sub returns -2
- wait on m_semaphore

Preemption

Thread 1:
- fetch_sub returned -1, so go signal m_semaphore
- thread 2 is unblocked and holds the semaphore

That's a bug. Semaphore was never signaled.

**Comment by rasendubi** [comment]
I believe there is a bug. Don't merge until discussed.

**Pull request merged**